pkg/statistics/handle/ddl/ddl.go

```diff
@@ -48,6 +50,23 @@ func NewDDLHandler(
 
 // HandleDDLEvent begins to process a ddl task.
 func (h *ddlHandlerImpl) HandleDDLEvent(t *util.DDLEvent) error {
+	sctx, err := h.statsHandler.SPool().Get()
+	if err != nil {
+		return err
+	}
+	defer h.statsHandler.SPool().Put(sctx)
+	if isSysDB, err := t.IsMemOrSysDB(sctx.(sessionctx.Context)); err != nil {
+		return err
+	} else if isSysDB {
+		// NOTE FOR EXCHANGE PARTITION EVENT: When exchanging a partition with a system table,
+		// we need to adjust the global statistics based on the count delta and modify count delta.
+		// However, due to the system table involved, a complete update of the global statistics isn't feasible.
+		// As a result, we bypass the statistics update for the table in this scenario.
```

Comment:
I see that when EXCHANGE PARTITION happens, we use the db of the normal table to do the check. So if someone creates a partitioned system table and exchanges it with a normal non-system table, the check here won't prevent it from creating an entry for the partitioned system table in `stats_meta`.

I think it's not a big problem but we better add some comments to avoid misunderstanding.

Reply:
I didn't mention it because it's so rare. I'll update the notes.

Reply:
I have added a case for it: https://github.com/pingcap/tidb/pull/52285/commits/489b9f92abf1a860598b7d20a1dd38830589f2cf

Reply:
I think the second line of the new comments is not correct.
It's not "does not have any effect". If there are stats for the original normal table, it will create stats_meta for the global stats (the table 104 below).

```
127.0.0.1:4000[test]> create table mysql.t1 (c1 int, c2 int) partition by range (c1) (partition p0 values less than (10));
Query OK, 0 rows affected (0.056 sec)

127.0.0.1:4000[test]> create table test.t2 (c1 int, c2 int);
Query OK, 0 rows affected (0.030 sec)

127.0.0.1:4000[test]> insert into t2 value(1,1);
Query OK, 1 row affected (0.005 sec)

127.0.0.1:4000[test]> show stats_meta;
+---------+------------+----------------+---------------------+--------------+-----------+-------------------+
| Db_name | Table_name | Partition_name | Update_time         | Modify_count | Row_count | Last_analyze_time |
+---------+------------+----------------+---------------------+--------------+-----------+-------------------+
| test    | t2         |                | 2024-04-03 15:22:12 |            1 |         1 | NULL              |
+---------+------------+----------------+---------------------+--------------+-----------+-------------------+
1 row in set (0.023 sec)

127.0.0.1:4000[test]> alter table mysql.t1 exchange partition p0 with table test.t2;
Query OK, 0 rows affected, 1 warning (0.048 sec)

127.0.0.1:4000[test]> show warnings;
+---------+------+---------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                               |
+---------+------+---------------------------------------------------------------------------------------+
| Warning | 1105 | after the exchange, please analyze related table of the exchange to update statistics |
+---------+------+---------------------------------------------------------------------------------------+
1 row in set (0.011 sec)

127.0.0.1:4000[test]> show stats_meta;
+---------+------------+----------------+---------------------+--------------+-----------+-------------------+
| Db_name | Table_name | Partition_name | Update_time         | Modify_count | Row_count | Last_analyze_time |
+---------+------------+----------------+---------------------+--------------+-----------+-------------------+
| mysql   | t1         | p0             | 2024-04-03 15:22:12 |            1 |         1 | NULL              |
+---------+------------+----------------+---------------------+--------------+-----------+-------------------+
1 row in set (0.003 sec)

127.0.0.1:4000[test]> select * from mysql.stats_meta;
+--------------------+----------+--------------+-------+----------+
| version            | table_id | modify_count | count | snapshot |
+--------------------+----------+--------------+-------+----------+
| 448824326858997760 |      107 |            1 |     1 |        0 |
| 448824350391926785 |      104 |            2 |     1 |        0 |
+--------------------+----------+--------------+-------+----------+
2 rows in set (0.003 sec)
```
And in the log:
```log
[2024/04/03 15:23:42.189 +08:00] [INFO] [exchange_partition.go:83] ["Update global stats after exchange partition"] [category=stats] [globalTableSchema=mysql] [globalTableID=104] [globalTableName=t1] [countDelta=1] [modifyCountDelta=2] [partitionID=105] [partitionName=p0] [partitionCount=0] [partitionModifyCount=0] [tableID=107] [tableName=t2] [tableCount=1] [tableModifyCount=1]
```

Reply:
Oh. That was a mistake.

Reply:
I dropped the test case.